src/table/use-column-widths.tsx

```diff
@@ -63,26 +66,78 @@ const WidthsContext = createContext<WidthsContext>({
+  containerRef: React.RefObject<HTMLElement>;
   children: React.ReactNode;
 }
 
-export function ColumnWidthsProvider({ visibleColumns, resizableColumns, children }: WidthProviderProps) {
-  const visibleColumnsRef = useRef<(PropertyKey | undefined)[] | null>(null);
-  const [columnWidths, setColumnWidths] = useState<Record<PropertyKey, number>>({});
+export function ColumnWidthsProvider({ visibleColumns, resizableColumns, containerRef, children }: WidthProviderProps) {
+  const visibleColumnsRef = useRef<PropertyKey[] | null>(null);
+  const containerWidthRef = useRef(0);
+  const [columnWidths, setColumnWidths] = useState<null | Record<PropertyKey, number>>(null);
 
   const cellsRef = useRef<Record<PropertyKey, HTMLElement>>({});
+  const stickyCellsRef = useRef<Record<PropertyKey, HTMLElement>>({});
   const getCell = (columnId: PropertyKey): null | HTMLElement => cellsRef.current[columnId] ?? null;
-  const setCell = (columnId: PropertyKey, node: null | HTMLElement) => {
+  const setCell = (sticky: boolean, columnId: PropertyKey, node: null | HTMLElement) => {
+    const ref = sticky ? stickyCellsRef : cellsRef;
     if (node) {
-      cellsRef.current[columnId] = node;
+      ref.current[columnId] = node;
     } else {
-      delete cellsRef.current[columnId];
+      delete ref.current[columnId];
     }
   };
 
+  const getColumnStyles = (sticky: boolean, columnId: PropertyKey): React.CSSProperties => {
+    const column = visibleColumns.find(column => column.id === columnId);
+    if (!column) {
+      return {};
+    }
+
+    if (sticky) {
+      return { width: cellsRef.current[column.id]?.offsetWidth || (columnWidths?.[column.id] ?? column.width) };
+    }
+
+    if (resizableColumns && columnWidths) {
+      const isLastColumn = column.id === visibleColumns[visibleColumns.length - 1]?.id;
+      const totalWidth = visibleColumns.reduce((sum, { id }) => sum + (columnWidths[id] || DEFAULT_COLUMN_WIDTH), 0);
+      if (isLastColumn && containerWidthRef.current > totalWidth) {
+        return { width: 'auto', minWidth: column?.minWidth };
+      } else {
+        return { width: columnWidths[column.id], minWidth: column?.minWidth };
+      }
+    }
+    return {
+      width: column.width,
+      minWidth: column.minWidth,
+      maxWidth: !resizableColumns ? column.maxWidth : undefined,
+    };
+  };
+
+  // Imperatively sets width style for a cell avoiding React state.
+  // This allows setting the style as soon container's size change is observed.
+  const updateColumnWidths = useStableCallback(() => {
+    for (const column of visibleColumns) {
+      setElementWidths(cellsRef.current[column.id], getColumnStyles(false, column.id));
+    }
+    // Sticky column widths must be synchronized once all real column widths are assigned.
+    if (Object.keys(stickyCellsRef.current).length > 0) {
+      for (const { id } of visibleColumns) {
+        setElementWidths(stickyCellsRef.current[id], getColumnStyles(true, id));
```

Comment:
Don't you want to add a check here that this column id is sticky, instead of relying on some code 3 levels down below to capture this with an undefined check?

Reply:
I don't get it. Do you suggest checking the `stickyCellsRef.current[id]` to be defined?

Reply:
At the very least, yes. But this is also not solid, because sticky is a dynamic preference, can change in runtime and I am not sure how much I can trust `stickyCellsRef` object.

Reading column definition sounds better

Reply:
So we do read visibleColumns to get the list of IDs to sync. The expectation is that the sticky header was already rendered and the references are available, however, the setElementWidths accepts an element or undefined as its first argument so no explicit check is needed. If we add an explicit check here there is nothing extra we can do if the ref is not available.

Reply:
So, "relying on some code 3 levels down" is your final word on this.

Clearly see being a source of a similar scale regressions as the past revert, but feel free to go ahead.

Reply:
Reminder about this thread. Still want to hear your plan how to deal with possible bugs in this code

Reply:
The code is already updated. The setElementWidths no longer accepts undefined as target.